### ISSUE TITLE
dataclass-ness of class not detected in argument to a decorator

### ISSUE THREAD

**Issue by tpgillam** [microsoft/pyright#10417]
Consider the following example, tested with pyright 1.1.400 and python 3.13.2.

```python
import dataclasses
import typing

if typing.TYPE_CHECKING:
    import _typeshed


def f[T: type[_typeshed.DataclassInstance]](cls: T) -> T:
    assert dataclasses.is_dataclass(cls)
    assert hasattr(cls, "__dataclass_fields__")
    return cls


@dataclasses.dataclass
class A:
    x: int

A1 = f(A)
assert A1 is A
typing.assert_type(A1, type[A])  # This is fine


@f  # Triggers pyright error (see below)
@dataclasses.dataclass
class B:
    x: int

print(B)
```

The example runs to completion without error in python.

Pyright reports exactly one issue:
```
  <snip>/moo.py:22:2 - error: Argument of type "type[B]" cannot be assigned to parameter "cls" of type "T@f" in function "f"
    Type "type[B]" is not assignable to type "type[DataclassInstance]"
      "B" is incompatible with protocol "DataclassInstance"
      Type "type[B]" is not assignable to type "type[DataclassInstance]"
        "__dataclass_fields__" is not present (reportArgumentType)
```

I think this is a false positive?

**Comment by erictraut** [comment]
Pyright is working as intended here. The `DataclassInstance` protocol defined in typeshed is relying on behaviors that are ambiguous and contested. You can read about it [here](https://github.com/microsoft/pyright/issues/4339). I recommend against using the `DataclassInstance` protocol, and I'd prefer that the typeshed maintainers remove or revise it. They currently have it commented as "experimental" and "might not work as expected for pyright". See [here](https://github.com/python/typeshed/blob/41eb1a6731aae4b429781704b09e9d0cc12379d0/stdlib/_typeshed/__init__.pyi#L348).
